Implement new multiplayer/playlists lounge screen designs

smoogipoo opened the pull request:
Closes https://github.com/ppy/osu/issues/9665

This is one step of the way towards implementing flyte's redesigns of this screen:
![image](https://user-images.githubusercontent.com/1329837/128186949-ebc94855-fd7e-4915-8176-4bcfa3452b3e.png)

This PR:
![image](https://user-images.githubusercontent.com/1329837/128188697-cfc9e37a-2562-43ea-8f3c-30a945b23966.png)

It isn't the full design because a lot is dependent on the rest of multiplayer being redesigned, which I'll get to. There's a few differences:
1. The background to the screen hasn't been updated. This is dependent on the match subscreen also being updated.
2. The multiplayer header (the one displaying `Multiplayer ● Lounge` hasn't been updated/removed. This currently results in a bit of empty space. This is dependent on the match subscreen also being updated.
3. I've moved the host text to below the title. The designs contain a description which we don't currently have.
4. The recent participants list only updates/populates when a room is selected. This will be fixed by https://github.com/ppy/osu-web/issues/7972.
5. The recent participants list will only display up to 50 users (3 + 47 hidden). This is an API-side limitation, but we'll probably need to think about changing the data structure rather than increasing, as it's probably not good to return 200+ users, for example, when only 3 are displayed.
6. The bottom bar/join button hasn't been implemented. I'm thinking to wait until that's done globally.

Along with redesigns, I've made some structural changes, such as:
- Moved `CreateNewRoom()` (which is also used for creating copies of rooms via right-click) to `LoungeSubScreen`.
- Moved the create room button to `LoungeSubScreen`.
- Reworked the lounge filter control's positioning/containering, because it was a bit haphazard and required manual offsets.

Closes #9665

smoogipoo committed:
Rename DrawableRoom test scene

smoogipoo committed:
Add clickability to test

smoogipoo committed:
Clean up selection box construction

smoogipoo committed:
Shorten room status messages

smoogipoo committed:
Redesign RoomStatusInfo

smoogipoo committed:
Initial redesign of DrawableRoom

smoogipoo committed:
Merge branch 'master' into lounge-redesign

smoogipoo committed:
Adjust some layout

smoogipoo committed:
Split pill display into abstract class

smoogipoo committed:
Add playlist count pill

smoogipoo committed:
Remove pill inheritance

smoogipoo committed:
Add rank range pill

smoogipoo committed:
Add host under room title

smoogipoo committed:
Add solid background

smoogipoo committed:
Add recent participants

smoogipoo committed:
Fix 1px gaps in DrawableRoom background

smoogipoo committed:
Remove RoomInspector from the lounge

smoogipoo committed:
Renamespace classes

smoogipoo committed:
wip

smoogipoo committed:
Merge branch 'master' into lounge-redesign

smoogipoo committed:
Add special category pill + secondary background

smoogipoo committed:
Remove spotlights glow

smoogipoo committed:
Add default background to panel, reduce nesting

smoogipoo committed:
Add shadow

smoogipoo committed:
Remove/fix paddings in lounge

smoogipoo committed:
Adjust layouts

smoogipoo committed:
Add filter background

smoogipoo committed:
Move create room button into the lounge

smoogipoo committed:
Merge branch 'master' into lounge-redesign

smoogipoo committed:
Merge branch 'master' into lounge-redesign

smoogipoo committed:
Fix more compile errors

smoogipoo committed:
Fix DrawableRoom test scene

smoogipoo committed:
Fix duplicate RoomManager

smoogipoo committed:
Increase background image resolution

smoogipoo committed:
Merge branch 'master' into lounge-redesign

smoogipoo committed:
Remove todo

peppy commented:
On a visual comparison of your screenshot to the design, the metrics seem a bit off (ie. the padding between the room_name and "hosted by" lines and the things above and below them). Is that kind of thing intentional/known or something that should be pointed out as part of the review?

Gonzalo-Bruna commented:
Is it showing the range of the star rating of the room? In that case, this PR should close my issue #9665, since there is no difficulties to hover anymore and now it shows the full range.

frenzibyte commented on osu.Game/Screens/OnlinePlay/Lounge/Components/PlaylistCountPill.cs:
```diff
@@ -0,0 +1,52 @@
+        {
+            count.Clear();
+            count.AddText(Playlist.Count.ToString(), s => s.Font = s.Font.With(weight: FontWeight.Bold));
+            count.AddText(" Maps");
```
Doesn't handle 1 map case, can use humanizer for that:
```csharp
            count.AddText(" Map".ToQuantity(Playlist.Count, ShowQuantityAs.None));
```

peppy replied:
Should also probably be "beatmap"

frenzibyte commented on osu.Game/Screens/OnlinePlay/Lounge/Components/DrawableRoom.cs:
```diff
@@ -112,114 +133,209 @@ public DrawableRoom(Room room)
+                                                Anchor = Anchor.CentreLeft,
+                                                Origin = Anchor.CentreLeft,
+                                            },
+                                            new StarRatingRangeDisplay
+                                            {
+                                                Anchor = Anchor.CentreLeft,
+                                                Origin = Anchor.CentreLeft,
+                                                Scale = new Vector2(0.85f)
+                                            }
```
Has it been considered how this should display when there's only one map? right now it displays as both min and max, which feels slightly weird:
![CleanShot 2021-08-05 at 02 39 00](https://user-images.githubusercontent.com/22781491/128269074-72e0a386-9cce-4379-9090-9af87498003d.png)

I think displaying as a singular star rating display would look better (should do the same for `min.StarRating == max.StarRating` in general).

frenzibyte reviewed (commented):
Partial review on the room design side, haven't looked at `LoungeSubScreen` yet.

peppy commented:
Some more purely visual things I noticed:

- In `TestSceneDrawableRoom` it requires two clicks to see the selected state. Not sure if out of scope but felt a bit weird:

![20210805 143434 (dotnet)](https://user-images.githubusercontent.com/191335/128296588-b2822101-3e48-4662-883e-df55ffd14c5a.gif)

- The snippets of the background on the left are scaled weirdly, causing pixelation:

![20210805 143506 (dotnet)](https://user-images.githubusercontent.com/191335/128296671-81121444-153b-482f-a047-bfbdc882c1cc.png)

- Padding near the selection border seems lacking (might need @arflyte input as it doesn't look to have been a consideration in the design). Feels like the border needs to be on the "outside" to make it work correctly with the current padding values, assuming they are correct.

![20210805 143632 (dotnet)](https://user-images.githubusercontent.com/191335/128296849-babc930d-d345-4a37-98d8-c0e150a8d74d.png)

Already mentioned in the OP, but we probably want to get rid of the "double header" part as a priority, it feels weird.

- The "create room" button feels potentially a bit small. @arflyte is there any reason it is at the top rather than in the bottom next to the "Join" button in your designs? I guess it's actually supposed to be a tab control style selector? May need to give it more height at least.

peppy commented on osu.Game/Screens/OnlinePlay/Lounge/Components/DrawableRoom.cs:
```diff
@@ -101,6 +103,25 @@ public bool MatchingFilter
 
+        private int numberOfAvatars = 3;
+
+        public int NumberOfAvatars
```
Is this actually going to be something dynamic in the design? Feels unnecessary if not.

smoogipoo replied:
In the designs it's variable - the in-match-subscreen has 3 avatars while the lounge has 7. I suppose this is because there's more space in the lounge whereas the in-match version has additional buttons.
The host text is also a bit misplaced which would otherwise further reduce the available width, but as I mentioned in the OP I've placed it where it is because we don't have a description and the details on the left would look weird.

peppy commented on osu.Game/Screens/OnlinePlay/Lounge/Components/DrawableRoom.cs:
```diff
@@ -112,114 +133,209 @@ public DrawableRoom(Room room)
+                    Name = @"Room content",
                     RelativeSizeAxes = Axes.Both,
-                    Padding = new MarginPadding(SELECTION_BORDER_WIDTH),
+                    Padding = new MarginPadding { Left = 20, Vertical = -0.5f },
```
Probably best to explain the negative padding inline for sanity.

peppy commented on osu.Game/Screens/OnlinePlay/Lounge/Components/DrawableRoom.cs:
```diff
@@ -112,114 +133,209 @@ public DrawableRoom(Room room)
+                                    new Box
+                                    {
+                                        RelativeSizeAxes = Axes.Both,
+                                        Colour = Color4Extensions.FromHex(@"#27302E"),
```
Same colour used in a few places, split out to a named variable (is this a special colour not in the scheme?)

smoogipoo replied:
Dunno where this colour comes from. It appears somewhere inbetween the blue and green colour schemes - it's more like a cyan hue?

@arflyte Where does the background colour of these multiplayer panels come from?
![image](https://user-images.githubusercontent.com/1329837/128349115-79626d6c-5292-4448-b2f7-2f5af01d136c.png)

frenzibyte replied:
FWIW it seems *somewhat* similar to `B5`'s HSL:
- this – `#27302E`: `167°, 10%, 17%`
- `B5` – `#222A28`: `160°, 10%, 15%`

peppy commented on osu.Game/Screens/OnlinePlay/Lounge/Components/EndDateInfo.cs:
```diff
@@ -0,0 +1,65 @@
+                var diffToNow = Date.Subtract(DateTimeOffset.Now);
+
+                if (diffToNow.TotalSeconds < -5)
+                    return $"Closed {base.Format()}";
```
Probably good practice to start localising newly added strings.

smoogipoo replied:
It's not a newly-added string, and the base component which also needs to change is used in a few more places. Will leave to a future PR.

peppy commented on osu.Game/Screens/OnlinePlay/Lounge/Components/RankRangePill.cs:
```diff
@@ -0,0 +1,77 @@
+            }
+
+            rankFlow.AddText("#");
+            rankFlow.AddText(Room.Users.Select(u => u.User?.Statistics.GlobalRank ?? 0).DefaultIfEmpty(0).Min().ToString(), s => s.Font = s.Font.With(weight: FontWeight.Bold));
```
Should probably be formatted with `0,#` or whatever we need to add thousands separators.

peppy commented:
I think I mentioned this in voice, but we also probably want to pull out the selected room polling logic. Maybe that's best left for a separate PR though (shall we open a tracking issue so it isn't forgotten?). I think it will still be required for playlists once a room is joined, but that can exist in the `PlaylistRoomSubScreen1.

smoogipoo committed:
Merge branch 'master' into lounge-redesign

smoogipoo committed:
Fix compilation failure

smoogipoo committed:
Adjust spacing

smoogipoo committed:
Fix pluralisation

smoogipoo committed:
Hide star rating max display for equal difficulties

smoogipoo committed:
Use higher res background image

smoogipoo committed:
Increase default number of avatars

smoogipoo committed:
Explain negative padding

smoogipoo committed:
Add constant for background colour

smoogipoo committed:
Add thousands separator to rank range pill

smoogipoo commented:
Please open a separate issue for removing selected room polling, yes.

smoogipoo committed:
Update recent participants list to use participant_count

frenzibyte commented on osu.Game/Screens/OnlinePlay/Lounge/Components/PlaylistCountPill.cs:
```diff
@@ -0,0 +1,54 @@
+        {
+            InternalChild = new PillContainer
+            {
+                Child = count = new OsuTextFlowContainer(s => s.Font = OsuFont.GetFont(size: 12))
+                {
+                    AutoSizeAxes = Axes.Both,
+                    Anchor = Anchor.Centre,
+                    Origin = Anchor.Centre,
+                }
```
Another thing I've missed to mention from my previous review is that currently centered text in pills get misaligned vertically:
![CleanShot 2021-08-05 at 17 44 45](https://user-images.githubusercontent.com/22781491/128369891-8cb65509-5af0-4b8e-ad81-6ba43f63543c.png)

In the figma designs, they're centered with a bottom padding of 1-2 depending on the font size:
![CleanShot 2021-08-05 at 17 46 05](https://user-images.githubusercontent.com/22781491/128370131-44082b0f-b9e7-43b0-a418-d16fa1b3f23a.png)

Should padding also be applied in the game, in a way that it becomes centered? or is that supposed to be handled somewhere framework-side?

Pointing this out because I also faced this during my work on the beatmap info overlay redesign, and had to apply `Padding` to not make the overlay pills look silly.

peppy replied:
Looks like the figma padding should be matched in this case.

smoogipoo committed:
Merge branch 'master' into lounge-redesign

smoogipoo committed:
Minor design adustments (paddings/sizing)

smoogipoo commented:
I also noticed that I implemented the rank range pill but haven't used it in the panel itself. Currently it's tied to multiplayer with the expectation that `DrawableRoom` would be derived to add multiplayer-specific things, however we could also add it to the lounge version if osu-web returned the relevant statistics somewhere, or otherwise it can only be added there once the multiplayer server is able to be queried from the lounge.

Not sure, any suggestions?

peppy commented:
> On a visual comparison of your screenshot to the design, the metrics seem a bit off (ie. the padding between the room_name and "hosted by" lines and the things above and below them - I think the whole rows are too tall maybe). Is that kind of thing intentional/known or something that should be pointed out as part of the review?

This still seems to be off. Can you confirm you saw this one?

smoogipoo commented:
Ah nope, I apparently missed that.

peppy commented:
I haven't gone through the code yet, but currently during fade out there's an overlapping duplicate of the beatmap background. Shouldn't there just be one instance with the gradient overlay applied over it (and then buffered)? Or is the left "tab" portion supposed to be from a different area of the background or something?

https://user-images.githubusercontent.com/191335/128489319-e28124d9-5b37-4cb7-8bc2-a9cd24eeb852.mp4

smoogipoo committed:
Adjust spacings and sizings of left-side details

smoogipoo committed:
Remove unused params

smoogipoo committed:
Remove unused using

smoogipoo committed:
Remove unnecessary duplicate background

smoogipoo committed:
Buffer the entire star rating range to fix overlapping alpha

smoogipoo commented:
Fixed the above two issues as well. I was hoping to not have to buffer the star rating range display in order to not lose clarity, but I suppose that's required.

peppy commented:
Just buffering the background and not the text shouldn't lose any clarity, right?

smoogipoo commented:
That's easier said than done. I suppose I could just reimplement `StarRatingDisplay` for this component, as right now it's just wrapping two of those and filling in the centre. May be best left to a separate PR...

peppy commented:
I wouldn't worry too much, can be done later (the whole buffering part if you'd prefer). There's many other places we have similar issues, and if I'm going to start nitpicking the lounge has more itself.

smoogipoo commented:
As for the background, it turns out that the top layer wasn't necessary at all, so I've just left the background + the padded gradient on top.

bdach commented on osu.Game/Screens/OnlinePlay/Components/StarRatingRangeDisplay.cs:
```diff
@@ -85,6 +89,7 @@ private void updateRange(object sender, NotifyCollectionChangedEventArgs e)
 
             minDisplay.Current.Value = minDifficulty;
             maxDisplay.Current.Value = maxDifficulty;
+            maxDisplay.Alpha = Precision.AlmostEquals(minDifficulty.Stars, maxDifficulty.Stars) ? 0 : 1;
```
This is using the default precision of `1e-7`, while star ratings are rounded to 2 decimal places - it may be the case that this will display ranges like [3.00*, 3.00*] if the underlying numbers are something like 2.997*-3.002*.

Setting the precision to `0.005` is not necessarily valid here either, though (consider ranges like 2.9901*-2.9951*). The way to go here would probably be to round both to 2 decimal places and compare that (probably directly at that point?)

bdach commented on osu.Game/Screens/OnlinePlay/Lounge/Components/RoomStatusPill.cs:
```diff
@@ -0,0 +1,67 @@
+
+        private void updateDisplay()
+        {
+            RoomStatus status = EndDate.Value < DateTimeOffset.Now ? new RoomStatusEnded() : Status.Value ?? new RoomStatusOpen();
```
This line is a bit much (ternary into null-coalesce). Would maybe split lines for the branches of the ternary or add some parens for clarification.

peppy replied:
Also not sure it should be allowed to be null in the first place. Maybe worthwhile to set a default value to avoid the need for the `??` logic?

smoogipoo replied:
It shouldn't be possible to be null. This may have been legacy code and `Room` defines a default value. Removed.

bdach commented on osu.Game/Screens/OnlinePlay/Lounge/Components/RoomStatusPill.cs:
```diff
@@ -0,0 +1,67 @@
+            pill.Background.FadeColour(status.GetAppropriateColour(colours), firstDisplay ? 0 : 100);
+            statusText.Text = status.Message;
+
+            firstDisplay = false;
```
Usually this is achieved with a `FinishTransforms()` call but I suppose it works.

peppy replied:
I also prefer `FinishTransforms` for this kind of thing, but I remember smoogi potentially being against that (as it has a larger scope than may be intended).

smoogipoo replied:
I dunno. I'm still against it, but in the interest of keeping things moving and not continuing the discussion I'll use `FinishTransforms()` until a point in time where it breaks and I re-assert my original implementation as the way it should be done.

bdach commented on osu.Game/Screens/OnlinePlay/Multiplayer/CreateMultiplayerMatchButton.cs:
```diff
@@ -22,6 +22,7 @@ public class CreateMultiplayerMatchButton : CreateRoomButton
         [BackgroundDependencyLoader]
         private void load()
         {
+            SpriteText.Font = SpriteText.Font.With(size: 14);
```
This is specified in both inheritors of `CreateRoomButton`. Possible to hoist this up there.

bdach reviewed (commented):
I've skimmed this diff some more. Mostly abstained on drilling down into the design bits since I presume those have been looked at enough.

peppy commented:
I'd prefer the buffering of star rating display was undone rather than buffering the whole thing (as previously mentioned, fixing this can be left for a later date). Doing it this way makes it pretty illegible:

![20210809 161554 (dotnet)](https://user-images.githubusercontent.com/191335/128671325-42c04e9e-029f-4d62-87f2-f8a3d8ed5ce1.png)

peppy committed:
Move font sizing to base class

peppy committed:
Split status retrieval into its own method

peppy committed:
Merge branch 'master' into lounge-redesign

peppy committed:
Remove `MaskingSmoothness` to avoid making sheered container blurry

smoogipoo committed:
Merge branch 'master' into lounge-redesign

smoogipoo committed:
Revert "Buffer the entire star rating range to fix overlapping alpha"

This reverts commit c680012523ff9f4a1e8ab10f788401a9f3b68986.

smoogipoo committed:
Round star ratings before comparing

smoogipoo committed:
Use FinishTransforms()

smoogipoo committed:
Don't use null-propagation for status

peppy committed:
Merge branch 'master' into lounge-redesign

peppy commented:
Doesn't necessarily need to be fixed here, but single pixel peeping issue:

![20210812 163555 (dotnet)](https://user-images.githubusercontent.com/191335/129156849-7f1bb34e-74aa-4fc1-8df6-3c72fb784d81.png)

peppy committed:
Fix padding around recent participants icon being uneven

peppy committed:
Add colour provider for multiplayer usage

peppy committed:
Fix incorrect colour for hidden user display

peppy committed:
Match font weight of design for hidden user count

peppy committed:
Fix recent participants hidden user logic not handling edge case correctly

Hiding just one user never makes sense, so this will now always show up
to the required circle count until two users are required to be hidden.
This will make the listing more consistent with the width requirement
spec.

peppy committed:
Simplify logic and test/fix edge case

peppy committed:
Add total user count

peppy committed:
Add some padding between scroll bar and content

peppy commented:
@smoogipoo i've fixed a few issues, please check through and make sure you are okay with them. the `RecentParticipantList` turned out to be a bit of a pain but hopefully should be more consistent in display as a result.

smoogipoo committed:
Add colour provider to online play dependencies

smoogipoo commented:
Seems okay.
